flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java

```diff
@@ -0,0 +1,172 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.source;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeutils.base.LongSerializer;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is the single (non-parallel) monitoring task which takes a {@link FlinkInputFormat},
+ * it is responsible for:
+ *
+ * <ol>
+ *     <li>Monitoring snapshots of the Iceberg table.</li>
+ *     <li>Creating the {@link FlinkInputSplit splits} corresponding to the incremental files</li>
+ *     <li>Assigning them to downstream tasks for further processing.</li>
+ * </ol>
+ *
+ * <p>The splits to be read are forwarded to the downstream {@link StreamingReaderOperator}
+ * which can have parallelism greater than one.
+ */
+public class StreamingMonitorFunction extends RichSourceFunction<FlinkInputSplit> implements CheckpointedFunction {
+
+  private static final long serialVersionUID = 1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingMonitorFunction.class);
+
+  private static final long DUMMY_START_SNAPSHOT_ID = -1;
+
+  private final TableLoader tableLoader;
+  private final Schema projectedSchema;
+  private final List<Expression> filterExpressions;
+  private final ScanContext options;
+
+  private volatile boolean isRunning = true;
+  private transient Object checkpointLock;
+  private transient Table table;
+  private transient ListState<Long> snapshotIdState;
+  private transient long startSnapshotId;
+
+  public StreamingMonitorFunction(TableLoader tableLoader, Schema projectedSchema, List<Expression> filterExpressions,
+                                  ScanContext options) {
+    Preconditions.checkArgument(
+        options.snapshotId() == null && options.asOfTimestamp() == null,
+        "The streaming reader does not support using snapshot-id or as-of-timestamp to select the table snapshot.");
```

Comment:
It would be good to include what _should _ be used to configure the stream. Looks like it should be `startSnapshotId` with no `endSnapshotId`.

Reply:
If use did not provide a `start-snapshot-id`, then this function will use the default `current` snapshot id as the `start-snapshot-id`.   So we are allowed to pass `null` `start-snapshot-id`  but are not allowed to pass non-null `end-snapshot-id`.

Reply:
we should probably also check this
```
    Preconditions.checkArgument(scanContext.asOfTimestamp() == null,
        "Can't set asOfTimestamp in ScanContext for continuous enumerator");
```

Reply:
The current condition `ctxt.snapshotId() == null && ctxt.asOfTimestamp() == null` will require both the snapshotId and asOfTimestamp  to be null.  So we don't have to do the extra `scanContext.asOfTimestamp() == null` check now.  But I prefer to use the separate `Preconditions.checkArgument` for snapshotId and asOfTimestamp, so that people could distingush the cause quickly.